docs: 0.13 update for `container` plugin guide

vvagaytsev opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first pull request, please read our contributor guidelines in the https://github.com/garden-io/garden/blob/main/CONTRIBUTING.md file.
2. Please label this pull request according to what type of issue you are addressing (see "What type of PR is this?" below)
3. Ensure you have added or run the appropriate tests for your PR.
4. If the PR is unfinished, add `WIP:` at the beginning of the title or use the Github Draft PR feature.
5. Please add at least two reviewers to the PR. Currently active maintainers are: @edvald, @thsig, @eysi09, @Orzelius and @vvagaytsev.
-->

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

shumailxyz commented on docs/other-plugins/container.md:
```diff
@@ -2,180 +2,233 @@
 ## Building images
 
-A bare minimum `container` module just specifies common required fields:
+A bare minimum `container Build` action just specifies common required fields:
```
```suggestion
Following is a bare minimum `container Build` action configuration using the required fields:
```

shumailxyz commented on docs/other-plugins/container.md:
````diff
@@ -2,180 +2,233 @@
 ```
 
-If you have a `Dockerfile` next to this file, this is enough to tell Garden to build it. You can also specify `dockerfile: <path-to-Dockerfile>` if you need to override the Dockerfile name. You might also want to explicitly [include or exclude](../using-garden/configuration-overview.md#includingexcluding-files-and-directories) files in the build context.
+If you have a `Dockerfile` next to this file, this is enough to tell Garden to build it. You can also
+specify `spec.dockerfile: <path-to-Dockerfile>` if you need to override the Dockerfile name.
````
```suggestion
If you have a `Dockerfile` next to this file, this is enough to tell Garden to build it. However, you can override the `Dockerfile` name or path by specifying `spec.dockerfile: <path-to-Dockerfile>`.
```

shumailxyz commented on docs/other-plugins/container.md:
```diff
@@ -2,180 +2,233 @@
 
-You can specify [build arguments](https://docs.docker.com/engine/reference/commandline/build/#set-build-time-variables---build-arg) using the [`buildArgs`](../reference/module-types/container.md#buildArgs) field. This can be quite handy, especially when e.g. referencing other modules such as build dependencies:
+You can specify
+[build arguments](https://docs.docker.com/engine/reference/commandline/build/#set-build-time-variables---build-arg)
```
```suggestion
[build arguments](https://docs.docker.com/engine/reference/commandline/build/#build-arg)
```

The link has been slightly changed to point to correct section.

shumailxyz commented on docs/other-plugins/container.md:
````diff
@@ -2,180 +2,233 @@
 ```
 
-Garden will also automatically set `GARDEN_ACTION_VERSION` as a build argument, so that you can reference the version of the action being built.
+Garden will also automatically set `GARDEN_ACTION_VERSION` as a build argument, so that you can reference the version of
+the action being built.
````
```suggestion
Additionally, Garden automatically sets `GARDEN_ACTION_VERSION` as a build argument, which you can use to reference the version of action being built.
```

shumailxyz commented on docs/other-plugins/container.md:
```diff
@@ -2,180 +2,233 @@
 ## Using remote images
 
-If you're not building the container image yourself and just need to deploy an external image, you can skip the Dockerfile and specify the `image` field:
+If you're not building the container image yourself and just need to deploy an external image, you do not need to
+define a `Build` action. You only need to define a `Deploy` action:
```
```suggestion
If you're not building the container image yourself and just need to deploy an external image, you can use the `Deploy` action:
```

I think this part `you do not need to define a Build action` can be skipped here.

shumailxyz commented on docs/other-plugins/container.md:
```diff
@@ -2,180 +2,233 @@
 {% hint style="warning" %}
-Note that if there is a _Dockerfile_ in the same directory as the module configuration, and you still don't want to build it, you have to tell Garden not to pick it up by setting `include: []` in your module configuration.
+Note that if there is a _Dockerfile_ in the same directory as the action configuration, and you still don't want to
+build it, you have to tell Garden not to pick it up by setting `include: []` in your action configuration.
```
by setting `include` or `exclude`? 🤔

vvagaytsev replied:
This section was removed in https://github.com/garden-io/garden/commit/64a85fe9b55051ff10838635a16b9a76e72f994d

See the [discussion above](https://github.com/garden-io/garden/pull/4282#discussion_r1196440424).

Orzelius commented on docs/other-plugins/container.md:
```diff
@@ -2,180 +2,233 @@
 
 {% hint style="warning" %}
-Note that if there is a _Dockerfile_ in the same directory as the module configuration, and you still don't want to build it, you have to tell Garden not to pick it up by setting `include: []` in your module configuration.
+Note that if there is a _Dockerfile_ in the same directory as the action configuration, and you still don't want to
+build it, you have to tell Garden not to pick it up by setting `include: []` in your action configuration.
```
I don't think this section is needed anymore as there's no build action to begin with.

vvagaytsev replied:
Good catch! Removed it in https://github.com/garden-io/garden/pull/4282/commits/64a85fe9b55051ff10838635a16b9a76e72f994d

Orzelius commented on docs/other-plugins/container.md:
````diff
@@ -2,180 +2,233 @@
 ```
 
-By default, we use the tag specified in the `container` module `image` field, if any. If none is set there, we default to the Garden module version.
+By default, we use the tag specified in the `container` action's `spec.publishId` field, if any. If none is set there,
+we default to the Garden `Build` action version.
````
```suggestion
By default, we use the tag specified in the `container` action's `spec.publishId` field. If none is set,
we default to the Garden `Build` action version.
```

shumailxyz commented on docs/other-plugins/container.md:
```diff
@@ -2,180 +2,233 @@
 ## Running tests
 
-You can define both tests and tasks as part of any container module. The two are configured in very similar ways, using the `tests` and `tasks` keys, respectively. Here, for example, is a configuration for two different test suites:
+For `container` action type you can define both `Test` and `Run` actions. The two are configured in a very similar way.
```
```suggestion
For `container` type, you can define both `Test` and `Run` actions. The two are configured in a very similar way.
```
IMO `container` action type part is bit confusing in reading when i read the full line.

Orzelius commented on docs/other-plugins/container.md:
```diff
@@ -2,180 +2,233 @@
+in [this guide](../k8s-plugins/action-types/container.md) and the full spec in
+our [reference docs](../reference/action-types/Deploy/container.md).
 
 ## Running tests
```
Maybe add a note that these create a new pod and are not executed in the deploy. That means that the dependency on the deploy is also not needed. Kubernetes-exec tests can be used to execute in the existing deployment.

vvagaytsev replied:
Dependency on `deploy.frontend` is necessary here. The integration tests do some calls to `frontend` endpoints, so the tests will fail if the `frontend` app is not deployed.

Verified that in [`demo-project`](https://github.com/garden-io/garden/tree/0.13/examples/demo-project) example. Tried a cold run
```
gdev run frontend-integ
```
with and without `deploy.frontend` dependency. Got an error in the latter case.

shumailxyz commented on docs/other-plugins/container.md:
```diff
@@ -2,180 +2,233 @@
+for `container Test` actions.
 
-## Running tasks
+## Running arbitrary scripts and jobs
```
```suggestion
## Running arbitrary workloads
```

shumailxyz commented on docs/other-plugins/container.md:
```diff
@@ -2,180 +2,233 @@
+## Running arbitrary scripts and jobs
 
-Tasks are defined very similarly to tests:
+`Run` actons are defined very similarly to `Test` actions:
```
```suggestion
To run arbitrary workloads, any scripts or jobs, you can use the `Run` action which is defined similarly to the `Test` action:
```

shumailxyz commented on docs/other-plugins/container.md:
```diff
@@ -2,180 +2,233 @@
-In this example, we define a `db-migrate` task that runs `rake db:migrate` (which is commonly used for database migrations, but you can run anything you like of course). The task has a dependency on `my-database`, so that Garden will make sure the database is up and running before running the migration task.
+In this example, we define a `db-migrate` `Run` action that executes `rake db:migrate` (which is commonly used for
+database migrations, but you can run anything you like of course). The action has a dependency on `my-database` `Deploy`
+action, so that Garden will make sure the database is up and running before running the migration job.
```
```suggestion
action, so that Garden will make sure the database is deployed before running the migration job.
```

shumailxyz commented on docs/other-plugins/container.md:
```diff
@@ -2,180 +2,233 @@
+and `Run` actions.
 
-One thing to note, is that tasks should in most cases be _idempotent_, meaning that running the same task multiple times should be safe.
+One thing to note, is that `Run` actions should in most cases be _idempotent_, meaning that running the same action
```
```suggestion
One thing to note, is that `Run` actions should in most cases be _idempotent_, meaning that running the same `Run` action
```

shumailxyz commented on docs/other-plugins/container.md:
```diff
@@ -2,180 +2,228 @@
 
-Here, we declare `my-image` as a dependency for the `my-service` Helm chart. In order for the Helm chart to be able to reference the built container image, we must provide the correct image name and version.
+Here we do not need to declare an explicit _build_ dependency on `my-image` like `dependencies: [ build.my-app ]`.
+Instead, we do it implicitly via the references to the `Build` action outputs in `spec.value.image`.
```
```suggestion
Instead, we do it implicitly via the references to the `Build` action outputs in `spec.values.image`.
```

Orzelius commented:
wow @shumailxyz that's some good quality review there!

shumailxyz commented:
Thanks @vvagaytsev. Good stuff with update! Just some nits here and there to improve the overall copy.

vvagaytsev commented:
@shumailxyz @Orzelius, thanks for the great review and comments! I've addressed all suggestions and replied to all comments. Please do one more review round :)

Orzelius reviewed (dismissed):
Maybe even a third pair of eyes would be of use here to get a fresh perspective on this? This doc is very important to garden in general so maybe we can summon someone?

Walther commented on docs/other-plugins/container.md:
```diff
@@ -2,180 +2,230 @@
 ## Building images
 
-A bare minimum `container` module just specifies common required fields:
+Following is a bare minimum `container Build` action configuration using the required fields:
```
nit: this sentence could perhaps be revised - the `container Build` feels a bit odd, considering it is in fixed-width (as if a command or piece of config or code) but we do not use that verbatim in the actual configuration

perhaps we could say e.g. "Following is a bare minimum `Build` action using the `container` type:" or something similar?

shumailxyz replied:
Agree with suggestion. also differentiates between action kind and `type`. Less confusing 👍

shumailxyz replied:
Quite an improvement already. Thanks :)

Walther commented on docs/other-plugins/container.md:
````diff
@@ -2,180 +2,230 @@
 ```
 
-Garden will also automatically set `GARDEN_ACTION_VERSION` as a build argument, so that you can reference the version of the action being built.
+Additionally, Garden automatically sets `GARDEN_ACTION_VERSION` as a build argument, which you can use to reference the
+version of action being built.
````
Nit: what is the end-user benefit / feature / capability given here? Maybe add some phrasing (or link to another doc) explaining what this gives you, or how you can use this

vvagaytsev replied:
I don't know tbh :)

Orzelius replied:
I think it's somewhat obvious: you know what's the version of your app in the build process and can use it internally to set versions, render docs or clear caches. I think adding a link to the docker doc is a quick win here https://docs.docker.com/engine/reference/commandline/build/#build-arg

vvagaytsev replied:
Please check https://github.com/garden-io/garden/pull/4282/commits/07da07b65ab6a0f9081d4264509ed44a675163b9

Walther commented on docs/other-plugins/container.md:
```diff
@@ -2,180 +2,230 @@
+## Deploying applications
 
-The Kubernetes plugins (local or remote) can deploy container modules. You'll find the relevant information in [this guide](../k8s-plugins/action-types/container.md) and the full spec in our [reference docs](../reference/module-types/container.md#services).
+The Kubernetes plugins (local or remote) can deploy `container Deploy` actions. You'll find the relevant information
```
Nit: this sentence is also perhaps a bit too internal-technical. Maybe this could be rephrased to something more user-oriented, e.g.
> After your software has been built, you might also want to deploy it somewhere. For this, check out our guide on deploying to Kubernetes using `Deploy` actions

Walther commented on docs/other-plugins/container.md:
```diff
@@ -2,180 +2,230 @@
+## Deploying applications
 
-The Kubernetes plugins (local or remote) can deploy container modules. You'll find the relevant information in [this guide](../k8s-plugins/action-types/container.md) and the full spec in our [reference docs](../reference/module-types/container.md#services).
+The Kubernetes plugins (local or remote) can execute `Deploy` actions of `container` type. You'll find the relevant
+information in [this guide](../k8s-plugins/action-types/container.md) and the full spec in
+our [reference docs](../reference/action-types/Deploy/container.md).
```
(repeating the comment as the text changed a bit before i submitted the comment, making it immediately outdated)

Nit: this sentence is also perhaps a bit too internal-technical. Maybe this could be rephrased to something more user-oriented, e.g.
> After your software has been built, you might also want to deploy it somewhere. For this, check out our guide on deploying to Kubernetes using Deploy actions

Walther commented on docs/other-plugins/container.md:
```diff
@@ -2,180 +2,230 @@
 ## Running tests
 
-You can define both tests and tasks as part of any container module. The two are configured in very similar ways, using the `tests` and `tasks` keys, respectively. Here, for example, is a configuration for two different test suites:
+For `container` type, you can define both `Test` and `Run` actions. The two are configured in a very similar way.
```
Here we mention both `Test` and `Run` actions and the fact they are configured in a similar way. However, we only give an example of a `Test` action.

Should provide an example of a `Run` action as well, or at least mention some differences between them to explain the motivation?

vvagaytsev replied:
Changed this, see https://github.com/garden-io/garden/pull/4282#discussion_r1196670393

Walther commented on docs/other-plugins/container.md:
```diff
@@ -2,180 +2,230 @@
 
-In this example, we define a `db-migrate` task that runs `rake db:migrate` (which is commonly used for database migrations, but you can run anything you like of course). The task has a dependency on `my-database`, so that Garden will make sure the database is up and running before running the migration task.
+In this example, we define a `db-migrate` `Run` action that executes `rake db:migrate` (which is commonly used for
+database migrations, but you can run anything you like of course). The action has a dependency on `my-database` `Deploy`
+action, so that Garden will make sure the database is deployed before running the migration job.
```
could we simplify this too, maybe
```suggestion
database migrations, but you can run anything you like of course). The action has a dependency on the `my-database` deployment, so that Garden will make sure the database is deployed before running the migration job.
```

Walther commented on docs/other-plugins/container.md:
```diff
@@ -2,180 +2,230 @@
+action, so that Garden will make sure the database is deployed before running the migration job.
 
-Unlike tests, tasks can also be dependencies for services and other tasks. For example, you might define another task or a service with `db-migrate` as a dependency, so that it only runs after the migrations have been executed.
+Since Garden `0.13` `Test` actions can also be dependencies for any other kinds of actions, e.g `Build`, `Deploy`
+and `Run` actions.
```
On another thought: do we need this sentence at all here? This is in the `Run` section but we are now explaining something about `Test` actions

vvagaytsev replied:
Removed this in https://github.com/garden-io/garden/pull/4282/commits/48869558458a375f64b7fcaf5aabfa0f9b773412

Walther commented on docs/other-plugins/container.md:
```diff
@@ -2,180 +2,230 @@
+for `Run` actions of `container` type.
 
-## Referencing from other modules
+## Referencing from other actions
```
re: previous comment, we could potentially move the sentence here somewhere. Any action kinds can depend on any other actions, etc

Walther commented on docs/other-plugins/container.md:
```diff
@@ -2,180 +2,230 @@
 ## Mounting volumes and Kubernetes ConfigMaps
 
-`container` services, tasks and tests can all mount volumes and Kubernetes Configmaps.
+_Runtime_ `container` actions (i.e. `Deploy`, `Run`, and `Test`) can all mount volumes and Kubernetes Configmaps.
```
Nit: is this sentence a bit too under-the-hood technical as well? Could we say something like

> Volumes and ConfigMaps can be mounted in all `Deploy`, `Run`, and `Test` actions of the `container` type.

vvagaytsev replied:
Changed this in https://github.com/garden-io/garden/pull/4282/commits/5159529d7e2685d7720a977f7d3e0d2d4ac86b09

Walther commented on docs/other-plugins/container.md:
```diff
@@ -2,180 +2,230 @@
+action, so that Garden will make sure the database is deployed before running the migration job.
 
-Unlike tests, tasks can also be dependencies for services and other tasks. For example, you might define another task or a service with `db-migrate` as a dependency, so that it only runs after the migrations have been executed.
+Since Garden `0.13` `Test` actions can also be dependencies for any other kinds of actions, e.g `Build`, `Deploy`
```
actually, see https://github.com/garden-io/garden/pull/4282#discussion_r1196621068

vvagaytsev commented on docs/other-plugins/container.md:
```diff
@@ -2,180 +2,230 @@
+## Running arbitrary workloads
 
-Tasks are defined very similarly to tests:
+To run arbitrary workloads, any scripts or jobs, you can use the `Run` action which is defined similarly to the `Test`
```
Changed this in https://github.com/garden-io/garden/pull/4282/commits/2e4b6eb4f403570857d7a281c551d8c768a6f00b

vvagaytsev commented:
@Walther thanks for the detailed comments! I've addressed those. Please, take a look again :) Feel free to resolve the discussions that no longer need any changes :)

vvagaytsev commented:
Ah, there are a few more open discussions, sorry :) Looking into those now.

Orzelius commented:
This is going to be the best doc ever after all this triage
![image](https://github.com/garden-io/garden/assets/33936483/c4c83613-31cc-4275-a8b2-d26ac42ef25b)

Walther reviewed (dismissed):
Big thanks to everyone for making this guide so good! ✨

vvagaytsev commented:
@Walther @Orzelius I've made some minor changes in https://github.com/garden-io/garden/pull/4282/commits/d2c655516db199bd2a9cdf53251d928a20a53538

Please take a look. Let's get this merged today :)
